editor/components/rich-text/index.js

```diff
@@ -789,12 +789,30 @@ export class RichText extends Component {
 						return;
 					}
 
-					const anchor = this.editor.dom.getParent( this.editor.selection.getNode(), 'a' );
-					if ( ! anchor ) {
-						this.removeFormat( 'link' );
-					}
-					const { value: href, ...params } = formatValue;
-					this.applyFormat( 'link', { href, ...params }, anchor );
+					let anchor;
+
+					// Create only one undo level when inserting and modifying the link
+					this.editor.undoManager.transact( () => {
+						// Use built-in TinyMCE command to insert the link
+						// This takes care of deleting any existing links within the selection
+						const { value: href, ...params } = formatValue;
+						this.editor.execCommand( 'mceInsertLink', false, {
+							href,
+							...params,
+							'data-wp-temp-link': 1,
```

Comment:
As far as I can tell, there's no way to get the node that `mceInsertLink` creates, so I borrowed this workaround from the classic editor.

https://github.com/WordPress/WordPress/blob/48077799d9d84b41d7ee1ea3a0054f0641813359/wp-includes/js/wplink.js#L396-L463